In [22]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np

# Part 1: Load the dataset

In [23]:
# Load the given datasets
categorical_df = pd.read_csv('chronic_kidney_disease_categorical.csv')
numerical_df = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Print the data
# print('Numerical Data: ',numerical_df)
print('Categorical Data: ', categorical_df)


Categorical Data:       unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  g

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

- Age (numerical)
    - Represents the age of a patient in years.
- Blood Pressure (numerical)
    - Indicates the blood pressure of the patient measured in millimeters of mercury (mm/Hg).
- Specific Gravity (nominal)
    - Refers to the specific gravity of urine, which represents concentration. The values can be (1.005, 1.010, 1.015, 1.020, 1.025).
- Albumin (nominal)
    - Refers to the levels of albumin in urine, which can indicate kidney health. The possible values are (0, 1, 2, 3, 4, 5).
- Sugar (nominal)
    - Indicates the presence of sugar in urine. Similar to albumin, this can indicate various health conditions. The values can be (0, 1, 2, 3, 4, 5).
- Red Blood Cells (nominal)
    - Indicates the presence of red blood cells in urine, with possible values of (normal, abnormal).
- Pus Cell (nominal)
    - Indicates the presence of pus cells in the urine. Values are (normal, abnormal).
- Pus Cell clumps (nominal)
    - Indicates whether pus cell clumps are present in the urine. Values are (present, not present).
- Bacteria (nominal)
    - Indicates the presence of bacteria in urine, with values of (present, not present).
- Blood Glucose Random (numerical)
    - Represents a random measurement of blood glucose levels in milligrams per deciliter (mg/dL).
- Blood Urea (numerical)
    - Indicates the blood urea level, measured in milligrams per deciliter (mg/dL).
- Serum Creatinine (numerical)
    - Represents the serum creatinine level, which is an indicator of kidney function, measured in milligrams per deciliter (mg/dL).
- Sodium (numerical)
    - Indicates the sodium level in blood, measured in milliequivalents per liter (mEq/L).
- Potassium (numerical)
    - Represents the potassium level in blood, also measured in milliequivalents per liter (mEq/L).
- Hemoglobin (numerical)
    - Indicates the hemoglobin level in grams per deciliter (g/dL) of blood.
- Packed Cell Volume (numerical)
    - Represents the proportion of blood volume that is occupied by red blood cells, typically expressed as a percentage.
- White Blood Cell Count (numerical)
    - Indicates the number of white blood cells per cubic millimeter (cells/cumm) of blood.
- Red Blood Cell Count (numerical)
    - Indicates the number of red blood cells in millions per cubic millimeter (millions/cmm) of blood.
- Hypertension (nominal)
    - Indicates whether the patient has hypertension, with possible values of (yes, no).
- Diabetes Mellitus (nominal)
    - Indicates the presence of diabetes mellitus, with possible values of (yes, no).
- Coronary Artery Disease (nominal)
    - Indicates whether the patient has coronary artery disease, with possible values of (yes, no).
- Appetite (nominal)
    - Refers to the patient's appetite level, with possible values of (good, poor).
- Pedal Edema (nominal)
    - Indicates whether there is swelling in the feet and ankles (edema), with possible values of (yes, no).
- Anemia (nominal)
    - Indicates whether the patient has anemia, with possible values of (yes, no).
- Class (nominal)
    - Represents the classification of the patient regarding chronic kidney disease, with possible values of (ckd, not ckd).
- Features and Labels
    - Features: All the attributes listed above (Age, Blood Pressure, Specific Gravity, etc.) are considered features. Features are the inputs used to predict an outcome.
    - Labels: The label in this dataset is "Class," which indicates the classification of chronic kidney disease (either 'ckd' for chronic kidney disease or 'not ckd' for patients without it). Labels are the outputs you want to predict.
- Data Types
    - Numerical Features:
        - These features contain quantitative data that can be measured. Examples include:
        - Age
        - Blood Glucose Random
        - Serum Creatinine
- Categorical Features:
    - These features contain qualitative data that represent categories. Examples include:
        - Specific Gravity
        - Red Blood Cells
        - Hypertension
- Examples of Categorical and Numerical Columns
    - Categorical Columns:
        - Specific Gravity (nominal)
        - Albumin (nominal)
        - Hypertension (nominal)
    - Numerical Columns:
        - Age (numerical)
        - Blood Pressure (numerical)
        - Blood Urea (numerical)


# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [24]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
numerical_dupes = numerical_df.duplicated(subset='unique_id').sum()
print(numerical_dupes)
# Drop these duplicate rows
numerical_df = numerical_df.drop_duplicates(subset='unique_id',keep='first')

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
categorical_dupes = categorical_df.duplicated(subset='unique_id').sum()
print(categorical_dupes)
categorical_df = categorical_df.drop_duplicates(subset='unique_id',keep='first')


30
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

In [25]:
unique_ids_df1 = set(numerical_df['unique_id'].unique())
unique_ids_df2 = set(categorical_df['unique_id'].unique())
only_in_df1 = unique_ids_df1.difference(unique_ids_df2)
only_in_df2 = unique_ids_df2.difference(unique_ids_df1)
print(len(only_in_df1), len(only_in_df2))

0 10


Answer: There are 10 unique id's that are in the 2nd df that are not in the 1st. If they weren't then the information in the row without the unique id would be filled with NaN.

In [26]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_df = pd.merge(numerical_df, categorical_df, on='unique_id', how='outer')

#Print the combined dataset
pd.set_option('display.max_columns', None)
print(merged_df)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0   

        wbcc  rbcc   al   su     rbc        pc         pcc          ba  htn  \
0     3800.0   2.1  3.0  0.0  no

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [27]:
# Calculate the percentage of rows that contain atleast one missing value
missing_rows = merged_df.isna().any(axis=1)
percentage = missing_rows.mean() * 100
# Print %
print(percentage)

# Drop these rows from the dataset
merged_df = merged_df.dropna()
# Print the Dataset
print(merged_df)

61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su     rbc        pc         pcc          ba  htn  \
0     3800.0  

## Part 3.4: Sort the dataset according to the Labels

In [28]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_df = merged_df.sort_values(by='Target').reset_index(drop=True)
# Print the dataset
print(sorted_df)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su       rbc        pc         pcc          ba  htn  \
0     3800.0   2.1  3.0  0.0  

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [29]:
# Write code here
# 
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

def encode(df, column):
    # Create dummy variables
    dummies = pd.get_dummies(df[column], prefix=column)
    
    # Join the new dummy columns to the original dataframe
    df = pd.concat([df, dummies], axis=1)
    
    # Drop the original column
    df.drop(column, axis=1, inplace=True)
    
    return df

for val in categorical_columns:
    sorted_df = encode(sorted_df, val)
# Print the dataset
print(sorted_df)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su  rbc_abnormal  rbc_normal  pc_abnormal  \
0     3800.0   2.1  3.0  0.0         Fal

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [30]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
def remove_outliers(df):
    numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns  # Identify numerical columns

    for col in numerical_cols:
        mean = df[col].mean()
        std_dev = df[col].std()

        # Calculate upper and lower bounds for outliers
        upper_limit = mean + 3 * std_dev
        lower_limit = mean - 3 * std_dev
        
        # Filter out the outliers
        df = df[(df[col] <= upper_limit) & (df[col] >= lower_limit)]
        
    return df

sorted_df = remove_outliers(sorted_df)
# Print the dataset
print(sorted_df)

     unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv  \
2       343710  61.0  80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0   
7       438182  73.0  80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0   
8       474407  21.0  90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
10      481293  60.0  60.0  288.0   36.0  1.7  130.0  3.0   7.9  25.0   
12      514721  56.0  90.0  129.0  107.0  6.7  131.0  4.8   9.1  29.0   
..         ...   ...   ...    ...    ...  ...    ...  ...   ...   ...   
148     486397  20.0  70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0   
149     203466  22.0  60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0   
150     482105  44.0  60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0   
151     228849  58.0  80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0   
152     996076  57.0  60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su  rbc_abnormal  rbc_normal  pc_abnormal  \
2     9200.0   3.2  2.0  0.0          True       Fal

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [31]:
# Normalize the all Numerical Attributes in the dataset.

# Normalize only numerical columns
#Should result in a number between 0 and 1
standardized_df = sorted_df.copy()
numeric_columns = sorted_df.select_dtypes(include=['float', 'int']).columns
standardized_df[numeric_columns] = (sorted_df[numeric_columns] - sorted_df[numeric_columns].mean()) / sorted_df[numeric_columns].std()




# Print the dataset
print(standardized_df)

     unique_id       age        bp       bgr        bu        sc       sod  \
2    -0.791645  0.865230  0.852933  1.493768  4.145545  2.060955 -1.045246   
7    -0.427251  1.628581  0.852933  3.611474  3.915669  2.605840 -0.493149   
8    -0.287525 -1.679273  1.955716 -0.253339  0.007782  0.348459 -2.885568   
10   -0.260965  0.801618 -1.352632  4.537970 -0.145468  0.348459 -1.965407   
12   -0.132028  0.547168  1.955716  0.329030  2.574727  4.240496 -1.781375   
..         ...       ...       ...       ...       ...       ...       ...   
148  -0.241278 -1.742885 -0.249849  0.170202  0.161033 -0.196426 -1.045246   
149  -1.332588 -1.615660 -1.352632 -0.518052 -0.835095 -0.040745 -0.493149   
150  -0.257833 -0.216183 -1.352632 -0.570995  0.237658 -0.585630 -0.493149   
151  -1.234682  0.674393  0.852933 -0.438638  0.390908 -0.040745 -0.125084   
152   1.724636  0.610780 -1.352632 -0.306281  0.352596 -0.040745  1.715238   

          pot      hemo       pcv      wbcc      rbcc        al

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:
All columns ending with _no: htn_no, dm_no, cad_no, pcc_notpresent, ba_notpresent, pe_no, ane_no, rbc_abnormal, pc_abnormal, pcc_notpresent, ba_notpresent, appet_poor, Target_notckd

In [32]:
#Remove that column
columns_to_drop = [col for col in standardized_df.columns if any(keyword in col for keyword in ['abnormal', 'notpresent', 'poor', '_no'])]
columns_to_drop.extend(['Target_notckd'])
df_cleaned = standardized_df.drop(columns=columns_to_drop)

# Print the dataset
print("Columns after removal:", df_cleaned.columns)

Columns after removal: Index(['unique_id', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo',
       'pcv', 'wbcc', 'rbcc', 'al', 'su', 'pcc_present', 'ba_present',
       'htn_yes', 'dm_yes', 'cad_yes', 'appet_good', 'pe_yes', 'ane_yes',
       'Target_ckd'],
      dtype='object')


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [33]:
# Export the dataframe to a new csv file
df_cleaned.to_csv('cleaned_dataset.csv', index=False)


# Export the dataframe to a new json file
df_cleaned.to_json('cleaned_dataset.json', orient='records')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [34]:

df = pd.read_csv('cleaned_dataset.csv')

# Print the top 15 rows of the DataFrame
print(df.head(15))


    unique_id       age        bp       bgr        bu        sc       sod  \
0   -0.791645  0.865230  0.852933  1.493768  4.145545  2.060955 -1.045246   
1   -0.427251  1.628581  0.852933  3.611474  3.915669  2.605840 -0.493149   
2   -0.287525 -1.679273  1.955716 -0.253339  0.007782  0.348459 -2.885568   
3   -0.260965  0.801618 -1.352632  4.537970 -0.145468  0.348459 -1.965407   
4   -0.132028  0.547168  1.955716  0.329030  2.574727  4.240496 -1.781375   
5   -0.010512 -0.088958 -1.352632  1.229055  2.000038  1.593911  0.058948   
6    1.663751  0.928843 -0.249849  0.143731  0.084407  0.348459 -0.861213   
7    1.569401  0.483555  0.852933  2.579092  1.272099  2.060955 -0.677181   
8    1.445135  1.056068 -1.352632  3.240875  0.697409  2.372318 -0.677181   
9    1.135594  1.501356 -1.352632  0.037846  3.264355  3.150726 -0.861213   
10   1.078623  1.501356 -0.249849  2.711449  1.616912  1.827433 -1.413310   
11   0.279188  0.738005  0.852933  4.935040 -0.183781  0.037096 -3.437665   

### Paste the markdown table here
|   unique_id |       age |       bp |      bgr |         bu |       sc |       sod |       pot |     hemo |      pcv |     wbcc |     rbcc |    al |       su | pcc_present | ba_present | htn_yes | dm_yes | cad_yes | appet_good | pe_yes | ane_yes | Target_ckd |
|-------------|-----------|----------|----------|-----------|----------|-----------|-----------|----------|----------|----------|----------|---------|----------|-------------|-----------|---------|--------|---------|------------|--------|---------|------------|
| -0.791645   | 0.865230  | 0.852933 | 1.493768 | 4.145545  | 2.060955 | -1.045246 | 1.348927  | -3.152899| -3.129155| 0.533228 | -2.486487| 1.4986  | -0.178902| False       | False     | True    | True   | True    | False      | True   | True    | True       |
| -0.427251   | 1.628581  | 0.852933 | 3.611474 | 3.915669  | 2.605840 | -0.493149 | 2.300236  | -1.864417| -1.758514| -0.345191| -1.081602| 1.4986  | -0.178902| False       | False     | True    | True   | True    | True       | False  | False   | True       |
| -0.287525   | -1.679273 | 1.955716 | -0.253339| 0.007782  | 0.348459 | -2.885568 | -1.346450 | -2.876796| -3.281448| 1.938699 | -1.592469| 3.36094 | -0.178902| True        | True      | False   | False  | False   | True       | False  | True    | True       |
| -0.260965   | 0.801618  | -1.352632| 4.537970 | -0.145468 | 0.348459 | -1.965407 | -2.139207 | -3.060865| -2.976862| 3.168486 | -2.741921| 2.42977 | 5.545973 | True        | False     | True    | False  | False   | False      | False  | True    | True       |
| -0.132028   | 0.547168  | 1.955716 | 0.329030 | 2.574727  | 4.240496 | -1.781375 | 0.714721  | -2.508658| -2.367688| -0.696559| -2.231053| 1.4986  | -0.178902| False       | False     | True    | False  | False   | True       | False  | False   | True       |
| -0.010512   | -0.088958 | -1.352632| 1.229055 | 2.000038  | 1.593911 | 0.058948  | -0.553692 | -2.186537| -2.519981| 2.904960 | -2.486487| 0.567432| -0.178902| False       | False     | True    | True   | False   | True       | False  | False   | True       |
| 1.663751    | 0.928843  | -0.249849| 0.143731 | 0.084407  | 0.348459 | -0.861213 | 0.556169  | -0.898055| -0.844753| -0.037745| -1.592469| 2.42977 | -0.178902| False       | False     | True    | True   | False   | True       | False  | False   | True       |
| 1.569401    | 0.483555  | 0.852933 | 2.579092 | 1.272099  | 2.060955 | -0.677181 | 0.873272  | -1.680348| -1.606220| -0.257349| -1.847903| 2.42977 | 5.545973 | True        | True      | True    | True   | False   | True       | True   | False   | True       |
| 1.445135    | 1.056068  | -1.352632| 3.240875 | 0.697409  | 2.372318 | -0.677181 | 1.666030  | -2.324589| -2.367688| -0.213428| -2.231053| 3.36094 | 5.545973 | False       | True      | True    | True   | False   | False      | True   | False   | True       |
| 1.135594    | 1.501356  | -1.352632| 0.037846 | 3.264355  | 3.150726 | -0.861213 | 0.873272  | -1.450262| -1.453927| 3.168486 | -1.081602| 3.36094 | -0.178902| False       | False     | True    | True   | False   | False      | True   | False   | True       |
| 1.078623    | 1.501356  | -0.249849| 2.711449 | 1.616912  | 1.827433 | -1.413310 | 0.080514  | -1.910434| -1.758514| -1.047927| -1.975620| 2.42977 | -0.178902| True        | True      | True    | True   | True    | True       | False  | False   | True       |
| 0.279188    | 0.738005  | 0.852933 | 4.935040 | -0.183781 | 0.037096 | -3.437665 | -1.346450 | -1.910434| -1.453927| 1.279884 | -1.081602| 0.567432| -0.178902| False       | False     | False   | True   | False   | False      | False  | False   | True       |
| 0.257453    | -2.633461 | -1.352632| -0.597466| 1.042223  | -0.196426| -1.045246 | 0.873272  | -2.140520| -2.215394| 3.827301 | -0.443018| 3.36094 | -0.178902| False       | True      | False   | False  | False   | False      | False  | False   | True       |
| -0.063374   | 0.801618  | 1.955716 | -0.306281| 0.505846  | 0.815503 | -0.861213 | 1.348927  | -1.588313| -1.758514| 1.104201 | -1.337036| 1.4986  | -0.178902| False       | False     | False   | False  | False   | True       | False  | False   | True       |
| -1.619611   | 0.038267  | 0.852933 | 0.434916 | 3.800731  | 5.641629 | -1.597342 | 1.824582  | -1.956451| -1.301633| -0.784401| -1.464752| 3.36094 | -0.178902| False       | True      | False   | True   | False   | True       | True   | False   | True       |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Instead we can work with smaller data sets, or instead leverage it to create code that will carry out the conversions we want.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [35]:
# Code to rename all the columns in the dataset
rename_mapping = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc_present': 'Pus Cell Clumps',
    'ba_present': 'Bacteria',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'htn_yes': 'Hypertension',
    'dm_yes': 'Diabetes Mellitus',
    'cad_yes': 'Coronary Artery Disease',
    'appet_good': 'Appetite',
    'pe_yes': 'Pedal Edema',
    'ane_yes': 'Anemia',
    'Target_ckd': 'Target'
}

# Renaming the columns
df_cleaned.rename(columns=rename_mapping, inplace=True)

# Displaying the DataFrame with renamed columns
print(df_cleaned)

     unique_id       Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
2    -0.791645  0.865230        0.852933              1.493768    4.145545   
7    -0.427251  1.628581        0.852933              3.611474    3.915669   
8    -0.287525 -1.679273        1.955716             -0.253339    0.007782   
10   -0.260965  0.801618       -1.352632              4.537970   -0.145468   
12   -0.132028  0.547168        1.955716              0.329030    2.574727   
..         ...       ...             ...                   ...         ...   
148  -0.241278 -1.742885       -0.249849              0.170202    0.161033   
149  -1.332588 -1.615660       -1.352632             -0.518052   -0.835095   
150  -0.257833 -0.216183       -1.352632             -0.570995    0.237658   
151  -1.234682  0.674393        0.852933             -0.438638    0.390908   
152   1.724636  0.610780       -1.352632             -0.306281    0.352596   

     Serum Creatinine    Sodium  Potassium  Hemoglobin  Packed 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [37]:
# Converted SQL to Pandas code
df_grouped = df_cleaned.groupby('Target').size().reset_index(name='count')
print(df_grouped)

   Target  count
0   False    111
1    True     17


It groups together the Positive cases of CKD and the negative cases of CKD and creates a data frame that tells us the count of positive ckd and negative ckd.